### PR TITLE
🐛 [Fix] validation_step

### LINKED ISSUE #133: Too slow and incorrect validation metrics
## Describe the bug
There are two issues within the validation step inside yolo/tools/solver.py/ValidateModel at Lines 47-62, containing the following code:
```
    def validation_step(self, batch, batch_idx):
        batch_size, images, targets, rev_tensor, img_paths = batch
        H, W = images.shape[2:]
        predicts = self.post_process(self.ema(images), image_size=[W, H])
        batch_metrics = self.metric(
            [to_metrics_format(predict) for predict in predicts], [to_metrics_format(target) for target in targets]
        )

        self.log_dict(
            {
                "map": batch_metrics["map"],
                "map_50": batch_metrics["map_50"],
            },
            batch_size=batch_size,
        )
        return predicts
```

1, The call to self.metric() (which is a torchmetrics.detection.MeanAveragePrecision instance) at every validation step calculates the results not only for the in step provided batch, but for **all** values currently in its accumulator (so for every example accumulated so far from all the previous steps), resulting validation becoming progressively slower after each step. If one has more than a couple hundred validation images in the dataset, it makes validation take extremely long.

2, The collate function of the data loader fills up the "targets" tensor with dummy zeroes (boxes) for the images, which have less ground truth boxes than the image with the most labels (boxes), to be able to batch all the labels in to a single tensor. These dummy values however are not filtered out and are still there, when calculating the validation metrics. This leads to incorrect (lower than should be) metric values, since torchmetrics takes these dummy values as "missed" labels (boxes).

## Proposed (and tested) solution
Change the above code snippet in Validate Model to:
```
    def validation_step(self, batch, batch_idx):
        batch_size, images, targets, rev_tensor, img_paths = batch
        H, W = images.shape[2:]
        predicts = self.post_process(self.ema(images), image_size=[W, H])
        self.metric.update([to_metrics_format(predict) for predict in predicts],
                           [to_metrics_format(target[target.sum(1) > 0]) for target in targets])
        return predicts
```

1, Change the self.metric() call to self.metric.update(), which only updates the accumulator with the new values, but does not calculate the results yet (which is done in "on_validation_epoch_end")
2, Drop the logging of per batch metric values (which were incorrect anyway, as they were cumulative values)
3, Filter out the dummy boxes from the per image "target" tensors before appending them to the metrics accumulator.

### PR DESCRIPTION
**Pull request by Adamusen**
Changed the self.metric() call to self.metric.update().
Removed the logging of per batch metric values.
Filtered out the dummy boxes from the per image "target" tensors before appending them to the metrics accumulator.

fixes #133

### CODE REVIEW

**Comment by henrytsui000** [comment]
Thank you for your pull request!

I may need to postpone merging this PR because my original plan was to display the estimated mAP in a progress bar. This would allow debuggers to quickly monitor the validation state without having to wait until the entire validation process is complete. However, I haven’t finished implementing that feature yet.

That said, I’ll definitely consider your PR since it significantly speeds up the process.

Best regards,
Henry Tsui

**Comment by Adamusen** [comment]
Hey,

On top of the speed problem (which is quite bad on its own with higher number of validation images), the pull request contains a fix of an additional problem as well, which leads to incorrect results, see issue #133:

`target[target.sum(1) > 0]`

Tip: Whenever I debug the train/validation epochs, I usually rewrite the "`__len__`" function of the dataset inside data_loader.py to return some small value instead of the real one, yielding train and validation epochs to only consist of a couple of batches :)

Best regards,
Adam

**Comment by henrytsui000** [comment]
Hi Adam,  

Thank you for pointing this out! Regarding the padding target issue, how about adding:  
```python
prediction = prediction[prediction[:, 0] != -1]
```  
after line 486:  
https://github.com/WongKinYiu/YOLO/blob/51f507f61e733b2a1ef993a5bc544f4e8b749f1d/yolo/utils/bounding_box_utils.py#L486-L488


I believe this approach is more straightforward for filtering out padding targets than using `target.sum(1) > 0`. Let me know your thoughts!  

Best regards,  
Henry Tsui

**Comment by Adamusen** [comment]
Hey,

Yes, your proposed solution should work just as well :)

I was considering filtering based on class IDs being -1 as well first, but then I decided to go with the sum instead to make the solution more robust and not dependent on the dataloader's collate_fn filling class IDs with -1 values (not to break the solution if it got removed in the future or something :))

Best regards,
Adam

**Comment by henrytsui000** [comment]
Hi Adam,

I’d prefer filtering by -1 because I think the padding class will consistently be -1. In contrast, shifting bounding boxes or certain data augmentations applied by users might more easily accidentally cause the sum of values to > 0, potentially leading to unintended filtering.

Feel free to update the code if you agree, or let me know, and I’ll make the update. It’s up to you!

Best regards,
Henry Tsui

**Comment by Adamusen** [comment]
Hey,

Since the collate_fn is applied after the data_augmentations this shouldn't be an issue, as such, I find both solutions equally valid, so I leave it up to you to decide whichever one you prefer :)

Best regards,
Adam
